### PR TITLE
Sequential write large file integration tests

### PR DESCRIPTION
**Pull request by Tulsishah**
### Description
Adding tests to write 500MB files sequentially with 20MB chunk size.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - Run all the tests for testbucket and mountedDirectory flags.(status:SUCCESS)
11. Unit tests - NA
12. Integration tests - Run all the tests for testbucket and mountedDirectory flags.(status:SUCCESS)

### CODE REVIEW

File: tools/integration_tests/util/operations/file_operations.go [hunk 1]
```diff
@@ -192,3 +194,38 @@ func ReadFileSequentially(filePath string, chunkSize int64) (content []byte, err
...
+	file, err := os.OpenFile(filePath, os.O_RDWR|syscall.O_DIRECT|os.O_CREATE, FilePermission_0600)
+	if err != nil {
+		log.Printf("Error in opening file:%v", err)
+	}
+
+	// Closing file at the end.
+	defer CloseFile(file)
+
+	for offset < fileSize {
+		// Get random 200MB or remaining filesize data into chunk.
```
Comment (raj-prince) [inline tools/integration_tests/util/operations/file_operations.go:211]: 200 MB?

File: tools/integration_tests/util/operations/file_operations.go [hunk 2]
```diff
@@ -192,3 +194,38 @@ func ReadFileSequentially(filePath string, chunkSize int64) (content []byte, err
...
+		chunkSize = int64(math.Min(float64(chunkSize), float64(fileSize-offset)))
```
Comment (raj-prince) [inline tools/integration_tests/util/operations/file_operations.go:212]: Why are we converting back and forth (int to float & float to int)?
Instead:
if ((fileSize - offset) < chunkSize) {
    chunkSize = (fileSize - offset);
}

File: tools/integration_tests/util/operations/file_operations.go [hunk 3]
```diff
@@ -192,3 +194,38 @@ func ReadFileSequentially(filePath string, chunkSize int64) (content []byte, err
 	}
 	return
 }
+
+func WriteFileSequentially(filePath string, fileSize int64, chunkSize int64) (err error) {
+	var numberOfBytes int
```
Comment (raj-prince) [inline tools/integration_tests/util/operations/file_operations.go:199]: Why do we need to declare the variable on the top? We should always initialise, where the variable will be used.

File: tools/integration_tests/util/operations/file_operations.go (hunk 1 above)
Comment (Tulsishah) [reply tools/integration_tests/util/operations/file_operations.go:211]: chunk size is 200MB so we are first getting chunk size data or the data remaining to write in file(<200MB)

File: tools/integration_tests/util/operations/file_operations.go [hunk 4]
```diff
@@ -192,3 +194,38 @@ func ReadFileSequentially(filePath string, chunkSize int64) (content []byte, err
...
+		if err != nil {
+			log.Fatalf("error while generating random string: %s", err)
+		}
+
+		// Writes random 200MB or remaining filesize data into file.
+		numberOfBytes, err = file.Write(chunk)
+		if err != nil {
+			return
+		}
+		if int64(numberOfBytes) != chunkSize {
```
Comment (raj-prince) [inline tools/integration_tests/util/operations/file_operations.go:224]: Ideally, here also we should throw error.

File: tools/integration_tests/write_large_files/write_one_large_file_sequentially_test.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	"path"
+	"testing"
+
+	"github.com/googlecloudplatform/gcsfuse/tools/integration_tests/util/operations"
+	"github.com/googlecloudplatform/gcsfuse/tools/integration_tests/util/setup"
+)
+
+const FiveHundredMB = 500 * 1024 * 1024
+const FiveHundredMBFile = "fiveHundredMBFile.txt"
+const chunkSize = 200 * 1024 * 1024
```
Comment (raj-prince) [inline tools/integration_tests/write_large_files/write_one_large_file_sequentially_test.go:28]: Please always declare const in capital case.

Comment (Tulsishah) [reply tools/integration_tests/write_large_files/write_one_large_file_sequentially_test.go:28]: You mean all capital or start should be capital?

File: tools/integration_tests/write_large_files/write_one_large_file_sequentially_test.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+		t.Errorf("Error in writing file: %v", err)
+	}
+
+	// Check if 500MB data written in the file.
+	fStat, err := os.Stat(filePath)
+	if err != nil {
+		t.Errorf("Error in stating file:%v", err)
+	}
+
+	if fStat.Size() != FiveHundredMB {
```
Comment (raj-prince) [inline tools/integration_tests/write_large_files/write_one_large_file_sequentially_test.go:48]: Is there any way to compare checksum to validate the written content? If possible we can fix the content written to the file. Currently we are writing random data so we can't keep a fixed checksum for that.

**Review by raj-prince — CHANGES_REQUESTED** [review]
Suggested few changes.

**Comment by raj-prince** [comment]
Also, is there any reason to choose the chunkSize as 200MB?

**Comment by Tulsishah** [comment]
> Also, is there any reason to choose the chunkSize as 200MB?

In the read tests we did the same.

Chunks of 200MB from gcs, so better that we have file size > (200 * 2.5) MiB so that we read atleast 2 times and 0.5 so that we know what happens when we don't require full chunk, so 500 MiB should also work.

File: tools/integration_tests/write_large_files/write_one_large_file_sequentially_test.go (hunk 6 above)
Comment (Tulsishah) [reply tools/integration_tests/write_large_files/write_one_large_file_sequentially_test.go:48]: We can compare two file content, but the problem is where will we keep the fix content of 500MB ? I think comparing filesize may be enough for this.

File: tools/integration_tests/util/operations/file_operations.go [hunk 7]
```diff
@@ -192,3 +193,41 @@ func ReadFileSequentially(filePath string, chunkSize int64) (content []byte, err
...
+		if (fileSize - offset) < chunkSize {
+			chunkSize = (fileSize - offset)
+		}
+		chunk := make([]byte, chunkSize)
+		_, err = rand.Read(chunk)
+		if err != nil {
+			log.Fatalf("error while generating random string: %s", err)
+		}
+
+		var numberOfBytes int
```
Comment (raj-prince) [inline tools/integration_tests/util/operations/file_operations.go:219]: Can't we do in the next line?
numberOfBytes, err **:=** file.Write(chunk)

File: tools/integration_tests/util/operations/file_operations.go (hunk 1 above)
Comment (raj-prince) [reply tools/integration_tests/util/operations/file_operations.go:209]: Okay, we can remove 200MB from the comment. As it's a generic method, only knows about chunkSize.

File: tools/integration_tests/write_large_files/write_one_large_file_sequentially_test.go (hunk 6 above)
Comment (raj-prince) [reply tools/integration_tests/write_large_files/write_one_large_file_sequentially_test.go:48]: Okay.

File: tools/integration_tests/util/operations/file_operations.go (hunk 7 above)
Comment (Tulsishah) [reply tools/integration_tests/util/operations/file_operations.go:219]: No it is giving error.
```
../util/operations/file_operations.go:224:4: result parameter err not in scope at return
	../util/operations/file_operations.go:222:18: inner declaration of var err error
```

File: tools/integration_tests/write_large_files/write_one_large_file_sequentially_test.go (hunk 5 above)
Comment (raj-prince) [reply tools/integration_tests/write_large_files/write_one_large_file_sequentially_test.go]: start should be capital. -> ChunkSize

Comment (Tulsishah) [reply tools/integration_tests/write_large_files/write_one_large_file_sequentially_test.go]: done

**Comment by raj-prince** [comment]
> > Also, is there any reason to choose the chunkSize as 200MB?
> 
> In the read tests we did the same.
> 
> Chunks of 200MB from gcs, so better that we have file size > (200 * 2.5) MiB so that we read atleast 2 times and 0.5 so that we know what happens when we don't require full chunk, so 500 MiB should also work.

For read, default value of sequential-read-size-mb is 200MB. That's why we keep 200MB as chunk size. Here, we can reduce the chunk size around 20MB. As here, it depends on the end user, how much data wants to write before sync call.

**Comment by Tulsishah** [comment]
> > > Also, is there any reason to choose the chunkSize as 200MB?
> > 
> > 
> > In the read tests we did the same.
> > Chunks of 200MB from gcs, so better that we have file size > (200 * 2.5) MiB so that we read atleast 2 times and 0.5 so that we know what happens when we don't require full chunk, so 500 MiB should also work.
> 
> For read, default value of sequential-read-size-mb is 200MB. That's why we keep 200MB as chunk size. Here, we can reduce the chunk size around 20MB. As here, it depends on the end user, how much data wants to write before sync call.

Done

**Review by raj-prince — APPROVED** [review]

**Pull request merged**